### DISCUSSION TITLE
Is it really useful to document Android voices?

### DISCUSSION THREAD

**Discussion by HadrienGardeur** [HadrienGardeur/web-speech-recommended-voices#29] [General]
Given the state of voice selection on Android, I'm really wondering if documenting them is helpful. Quick reminder of the current issues with Android voice selection:

- #12 
- #14 
- https://github.com/HadrienGardeur/read-aloud-best-practices/issues/6

To make things worse, these voice names are also localized, which means that the more we extend our language support in this project, the more we need to document additional `altNames`.

The irony of it all is that we can actually document these Android voices better in Chrome OS than in Android itself:
https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/1d2476209f6c9f1b08cacf15036baa230ee9c453/json/en.json#L1243-L1259

I'm at a point where I'm wondering if I should give up on documenting these voices, as it really feels pointless.

If I had to summarize a list of pros/cons for having them in the list of recommended voices:

**Pros:**
* Slightly better voice name than what Android provides
* Implementers can rely strictly on recommended voices to offer a decent selector on all platforms

**Cons**
* Due to the bugs that I've documented, most of these voices won't work anyway
* Additional work whenever we add another language, for very little in return

I'm tempted to keep the status quo until we turn this repo into code. With code we could either:

* handle these `altNames` variations programmatically with locales
* and/or return other voices where there is no match for recommended voices

Any thoughts from anyone else on this? cc @chocolatkey @mickael-menu @danielweck

**Comment by Falcosc** [comment]
Thanks for all the work! You are amazing.

I prefer code as well, then you don't have to deal with handling how to enroll breaking interface changes if the old structure does not work anymore.

A function which returns voices in your recommended order would be amazing.
Currently, I concat recommendations of 2 of your language lists (user locale + English) to order the voices without removing any of them.

Result order looks like this:
1. User language voices in recommended order
2. English voices in recommend order
3. all other language and not detected voices

To give you some feedback, how I use your data: I reduce your structure to flat like this:
```
name
altName
altName
altName
name
altName
altName
altName
```
And then I order the available voices based on the index of name match in the flat list.

There is one disadvantage for this approach: high quality Apple voices are maybe not on top if the first installed high quality voice is for example "Lee (Premium)" since Lee is the last apple voice in your list.

**Reply by HadrienGardeur** [reply]
> Result order looks like this:
>
>1. User language voices in recommended order
>2. English voices in recommend order
>3. all other language and not detected voices

I'm also working on best practices in a separate repo, in case you haven't seen it yet: https://github.com/HadrienGardeur/read-aloud-best-practices/blob/main/voice-selection.md

If I had to summarize things:

- Instead of displaying a list of voices with all languages, it's best to show a language selector first, based on available languages on the device.
- Once a language has been selected (or inferred based on the base language of the publication), I would recommend grouping voices by regions (in practice, these are mostly countrie) and showing the region of the user first. If it can't be detected, try to infer the region from another language or fallback to the default region for that language.
- For each region, show recommended voices first and then everything else for that region.
- In the future, I'll also provide a list of voices to filter out (for example novelty voices on Apple devices).

> There is one disadvantage for this approach: high quality Apple voices are maybe not on top if the first installed high quality voice is for example "Lee (Premium)" since Lee is the last apple voice in your list.

It's fairly hard to detect high quality Apple voices because the voice name doesn't give you much information. I wouldn't flatten the list though, because you don't want to display a voice multiple times if several quality variants have been installed.

**Reply by Falcosc** [reply]
I know, but with the incorrect order of having Lee last, even if it's premium, combined with the iOS issue, I can't care less about a broken experience on Apple ;)

**Reply by HadrienGardeur** [reply]
When you say incorrect order for Lee, what do you mean exactly?

Here's the configuration for Lee:
https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/1d2476209f6c9f1b08cacf15036baa230ee9c453/json/en.json#L958-L1001

I've ordered things in a way that should ideally match the highest quality settings, but that's dependent on browser and locale:
- since Lee doesn't come preloaded, if you install a single variant it'll match with just "Lee" on Chromium-based browsers, with no other mean to detect which variant
- if you install more than one variant, Chromium-based browsers will use the "Lee (English (Australian))" pattern instead, with no easy way to identify which is which
- in Firefox, the variants are detected, which is why I've ordered them with the highest quality at the top (Lee has three quality variants)

This isn't perfect, but should work in most cases.

I'll open a separate discussion about Apple voices, as I think we could do something better if we listed their quality settings in those JSON files and handle localization with code.

**Reply by Falcosc** [reply]
If only lee is premium, all other voices without premium are ranked higher because lee is the last voice in the list, doesn't matter if we have the perfect lee installed, its still worse then the basic version of any other voice since lee is at the end of apple voices.

It will display lee correctly as premium but all non premium voices are above lee.


1. Karen (Australia)
2. Lee (Premium)

**Reply by HadrienGardeur** [reply]
I see what you mean but that's not very likely to happen:

- pre-loaded voices are all listed below the ones that you have to install: in the case of Australian voices, the preloaded one is Karen and it's line 1229: https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/1d2476209f6c9f1b08cacf15036baa230ee9c453/json/en.json#L1229-L1230
- the higher quality version of Karen is listed line 915: https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/1d2476209f6c9f1b08cacf15036baa230ee9c453/json/en.json#L915-L916
- there's also an `overrides` property that you can use for that, it tells you which voice to skip if you have any match for the higher quality variants of Karen: https://github.com/HadrienGardeur/web-speech-recommended-voices/blob/1d2476209f6c9f1b08cacf15036baa230ee9c453/json/en.json#L947

Even with that approach, it's possible to encounter cases where a higher quality variant is ranked lower than a medium quality one. But that's only the case if the user decided to install both voices, which isn't as much likely to happen. When you start installing additional voices, there's very little reason not to go for the best ones aside from the fact that it uses more space on your hard drive.

This could be solved with a more sophisticated approach though, where the quality variants for Apple are identified using a `quality` property (an array of strings), localization is documented and code used to filter/order things better.
I think this project will go down that road, I'll open a discussion about it later this week. But in the meantime, I'd say that it's already quite usable as-is and better than not filtering a voice list at all.

**Reply by Falcosc** [reply]
Now I see it. Wasn't aware that most of the apply voices have 2 entries because Lee is an exception with only one entry.
I thought overrides paramter was just another altName, but no, there is a whole other entry much lower in the quality index.

By the way, I stop supporting iOS. Speak API is behaiving different there, for example the voice is not working if the first interaction was triggered via promise or websocked or something.

Without MacOS you can't debug this stuff, since I can't connect to an iPhone to debug what's happaning.
And with the low quality iOS Voices it's just not worth it for me. And even worse are the unavoidable power management topics which turn off your javascript when the screen goes out.

Don't know what to do with MacOS because I don't know how to test things there. I did not expect any requirement to test things there, because Firefox and Chrome did behaive pretty identically for me. But after seeing the JS issues on iOS I guess there will be more on MacOS.

**Reply by HadrienGardeur** [reply]
> Now I see it. Wasn't aware that most of the apply voices have 2 entries because Lee is an exception with only one entry.
I thought overrides paramter was just another altName, but no, there is a whole other entry much lower in the quality index.

Most voices have a single entry, only preloaded voices have two entries.

> By the way, I stop supporting iOS. Speak API is behaiving different there, for example the voice is not working if the first interaction was triggered via promise or websocked or something.

Have you tried libraries that attempt to fix some of the existing issues with the implementation of WebSpeech API across browsers? For example https://jankapunkt.github.io/easy-speech/

Fingers crossed that iOS/iPadOS 18 will fix some of the worst bugs for voice selection, I guess we'll know very soon now that the developer beta is available.

**Reply by Falcosc** [reply]
I guess it will not fix this one:
https://communities.apple.com/de/thread/254066932?answerId=257918442622&sortBy=best#257918442622

It's just too much effort to work on iOS with all these issues.

I will revisit iOS in 1-2 Years when the EU law did give us enough developers complaining about the apple implementation of TTS. Every week I discover a new iOS problem.

Thank you

**Reply by Falcosc** [reply]
Microsoft is not much better, feels like a competition who has the most obvious bugs ;)
Multilingual Voices from Edge stop processing after reading "&" for all German multilingual voices.
It gets correctly read, but it throws an exception after that.
I don't know what's worse, reading Homograph between the words and inability to understand phrases between german words or having TTS skip the rest of the text on & 😅

Luckily, the other edge voices are still better than iOS and Firefox. It is even possible to barely understand English words without multilingual support. On other browsers, it's often impossible if you don't have the text as a clue of how to interpret the mess.
So yes, for German, Edge is the only solution so far if there is some internet slang involved.

**Reply by HadrienGardeur** [reply]
>I will revisit iOS in 1-2 Years when the EU law did give us enough developers complaining about the apple implementation of TTS. Every week I discover a new iOS problem.

One of my goal with this project is to document these issues. There are many bugs and limitations that get in the way of implementing a decent read aloud experience, but they're not centralized somewhere.

I've started a Github project specifically for that at: https://github.com/users/HadrienGardeur/projects/2

If you encounter any bug, feel free to either open an issue on this repo or [the main one dedicated to the project](https://github.com/HadrienGardeur/read-aloud-best-practices) (I'll triage and tag them accordingly). You can also reach out directly by email if that's better for you.

I'll make sure that several organizations working on accessibility are made aware of them, and hopefully this will give us enough leverage to prioritize some of these bugs at Apple/Google/Microsoft/Mozilla.

> Multilingual Voices from Edge stop processing after reading "&" for all German multilingual voices.

Yikes. Hopefully they can solve that issue on the server side, these multilingual voices are brand new and might be more prone to errors than others.

This is a good example of the kind of bugs that need to be files in the project dedicated to known-issues.

**Reply by Falcosc** [reply]
I have created 3 issues to go into more details. My other iOS issues are not related to Speech. But the "you need a Mac for web inspection" was the last straw before I asked by self "why supporting at all?" if the experience sucks anyway for them because of the voice quality.

**Reply by HadrienGardeur** [reply]
Thanks! I'll triage these issues tomorrow so don't be surprised if I move and/or rename them.

**Comment by Falcosc** [comment]
One more reason to find a more generic solution with fuzzy detection for Android Voices is the differences between iOS and Android.

On Android, you can have JavaScript Running while the display is off. This is interesting for any complex, long-running use case of Text to Speech.
On iOS you need an App to do this. Maybe Apple don't want to work on adding an exception for special JS declarations because they don't like if people are able to bypass the need of the App store by making browsers too powerful. For example audio streaming is a big market where apple wants to have a share, so quite convenient that streaming providers can not create iOS web apps for there platform ;)

On the other hand, users of long-running applications are more likely to spend more time with trying out voices and don't need recommendations as badly as website visitors.

**Reply by HadrienGardeur** [reply]
> On the other hand, users of long-running applications are more likely to spend more time with trying out voices and don't need recommendations as badly as website visitors.

Well the bad news is that Android makes this nearly impossible due to its current approach for voice selection (see issues listed in my initial message). It's better if you're building a native app, but not that great either since the API doesn't return a first name or a gender for voices.

> On Android, you can have JavaScript Running while the display is off. This is interesting for any complex, long-running use case of Text to Speech.
> On iOS you need an App to do this.

Apple doesn't like having anything running in the background on iOS/iPadOS, even with native apps. You usually have to work extra hard to do this and/or use specialized APIs.

**Reply by Falcosc** [reply]
Cool, Edge on Android is even more broken. I told you it's a challenge between Apple and Microsoft who has the most broken Browser TTS. I thought the Edge bug in your tracker is something new or only for a specific thing, no does not work at all since a long time.

Can you remind me what year we have? Maybe we should skip TTS and directly generate a Video based on the Text prompt and use that to hear the text, that seems to be easier than TTS in browser ;)

**Reply by Falcosc** [reply]
Edge Android uses the Edge Windows strings.

Since Edge Android issues are bugs while Chrome issues feel more like features I feel we get a solution on Edge first if at all.
Chrome voices are not worth the hassle, I would skip then until they get competitive. Maybe it makes more sense to have a look on other android browsers with much less market share firstwhile waitung for the Edge fix.
Meanwhile I use the Edge get voice list workarround. Thanks for your work.

**Comment by HadrienGardeur** [comment]
Now that we're truly implementing this work beyond a demo, I think that I can provide an answer to the initial question: "Is it really useful to document Android voices?".

And the answer is… not really. I'll continue to document one localized name (in the target language) and to indicate with `localizedName` that these voices are localized using Android's pattern for voices: `language Region` where language can be downcased or capitalized depending on capitalization rules for that language.

Going beyond that seems unnecessary because:

- these values are not truly voices ([see relevant issue](https://github.com/HadrienGardeur/web-speech-recommended-voices/issues/12))
- localization can be handled using `Intl.DisplayNames`
- and worst-case scenario they will be displayed as-is if we don't filter out non-recommended languages

I'm closing the discussion for now.
